### ISSUE TITLE
Still getting NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE with Objects.requireNonNull

### ISSUE THREAD

**Issue by cristalp** [spotbugs/spotbugs#2965]
I have read the issues regarding `Objects.requireNonNull`, so I expected the following to show no errors (using Spotbugs Maven plugin 4.8.4.0 with Spotbugs 4.8.4).

```
package foobar;

import java.util.Objects;
import java.util.function.Supplier;

import javax.annotation.CheckForNull;
import javax.annotation.Nullable;

public class SpotBugsTest {

  public void foo() {
    final String foo = getString("foo");
    Objects.requireNonNull(foo);
  }

  public void bar() {
    final String bar = getString("bar");
    Objects.requireNonNull(bar, "Bar must not be null");
  }

  public void baz() {
    final Supplier<String> supplier = ()-> "Baz must not be null";
    final String baz = getString("baz");
    Objects.requireNonNull(baz, supplier);
  }

  @CheckForNull
  private String getString(@Nullable final String text) {
    return text == null ? null : text;
  }
}
```

However, I get Spotbugs errors for all three cases.

```
    <BugInstance type='NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE' priority='Normal' category='STYLE' message='Possible null pointer dereference in ch.ipi.schutztitel.bo.SpotBugsTest.bar() due to return value of called method' lineNumber='18' />
    <BugInstance type='NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE' priority='Normal' category='STYLE' message='Possible null pointer dereference in ch.ipi.schutztitel.bo.SpotBugsTest.baz() due to return value of called method' lineNumber='24' />
    <BugInstance type='NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE' priority='Normal' category='STYLE' message='Possible null pointer dereference in ch.ipi.schutztitel.bo.SpotBugsTest.foo() due to return value of called method' lineNumber='13' />
```

I thought this was fixed with the issues I read. Am I doing something wrong?

P.S. Thanks a lot for SpotBugs and carrying the load of work!

**Comment by gtoison** [comment]
My understanding is that the methods are analysed separately, so when analyzing your `foo`, `bar` or `baz` methods SpotBugs is not smart enough to see that the calls to `getString("xyz")` won't return null. At the same time your code says that it _can_ return null with `@CheckForNull`.

There recent change in #2709 was to let SpotBugs know that `foo` won't be null _after_ `Objects.requireNonNull(foo);`

**Comment by ryanrupp** [comment]
#2709 looks like it should be fixed in 4.8.4 (which the reporter is using). I'm running into this too and was thinking an upgrade to pull in #2709 would resolve this but seeing the same. It seems from the error it's actually complaining about the use of `Objects.requireNonNull` - it's complaining that you have a method that could return null and are now passing a potential null to `Objects.requireNonNull` - but that's also sort of the point of the call/check (to fail and then make any downstream code understand it would be non-null by the runtime check).

FWIW my usage is slightly different but the same idea, I have a method that the return is annotated with `org.springframework.lang.Nullable` and pass this to `Objects.requireNonNull`. As a workaround oddly using Spring's `Assert.notNull` works i.e. doesn't like this:

```
MyType nullableResult = callNullableMethod();
nullableResult = requireNonNull(nullableResult);
```

but is okay with this:

```
MyType nullableResult = callNullableMethod();
Assert.notNull(nullableResult);
```

Note that `Assert.notNull` object argument is marked with `org.springframework.lang.Nullable` as well so maybe that's why this works?

I also tried Guava's `Preconditions.checkNotNull` which is annotated with `@CheckForNull` (not sure if that matters) - this failed the same way as `Objects.requireNonNull` though.

I started looking at the spotbugs code to see possibly what it's doing with `Preconditions` and `Objects` and I see for Preconditions [here](https://github.com/spotbugs/spotbugs/blob/adcbefe732c12e4e1735349a27d35e3be8a8089a/spotbugs/src/main/java/edu/umd/cs/findbugs/ba/DefaultNullnessAnnotations.java#L408) and Objects [here](https://github.com/spotbugs/spotbugs/blob/adcbefe732c12e4e1735349a27d35e3be8a8089a/spotbugs/src/main/java/edu/umd/cs/findbugs/ba/DefaultNullnessAnnotations.java#L425) - if I'm reading this write it's marking both of those parameters as non-null. I'm trying to understand if this is correct, I could see maybe this a design choice that static code probably shouldn't be passing something that's known to be potentially null to these... but at the same time this is how we're trying to say "trust me I know this won't be null". The commit for this is very old though [here](https://github.com/spotbugs/spotbugs/commit/a9cd2eb80193f094705830b20d8c78b9a13cc409#diff-c4d76590b2f3083d9f92678f7170378958cfdf5273671b7654b31dc7d536a804R389) and then `Objects.requireNonNull` just follows that pattern in later commit [here](https://github.com/spotbugs/spotbugs/commit/cc555d358001be374e9fb2a0d3081430ce6a0fd1#diff-c4d76590b2f3083d9f92678f7170378958cfdf5273671b7654b31dc7d536a804R418). 

That said, need to look more into how the unit test [here](https://github.com/spotbugs/spotbugs/pull/2709/files#diff-8a2f521e27edc89ce0fa7c448ab83ea426cd34640e42defc7fce4f9c28adc715R14) wouldn't fail then.

**Comment by ryanrupp** [comment]
My issue seemed tied to multi-line handling so I logged https://github.com/spotbugs/spotbugs/issues/3186 with some example test cases.

However, this one also seems similar/maybe the same fix or at least the same area of code. It's struggling on the same area of code at least where it's doing line comparisons and in the multi-statement example ends up skipping the later half of the logic in `FindNullDeref.callToAssertionMethod` which is the logic to determine that this is a not an issue. It's early breaking [here](https://github.com/spotbugs/spotbugs/blob/master/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java#L1361).

Example [here](https://github.com/ryanrupp/spotbugs/commit/b84c649992260b9baa52c7b6f568f03b2efb2e5b#diff-8a2f521e27edc89ce0fa7c448ab83ea426cd34640e42defc7fce4f9c28adc715R58)

**Comment by Wolf2323** [comment]
We have a similar issue I think it is related, here is my code snippet:

```java
			final InventoryOpenCloseInfo info = INFO_MAP.remove(player);
			if (info == null) {
				return new AnalysisResult<>(getName(), null, event, AnalysisResult.Result.SKIPPED);
			}
			info.inventoryCloses(inventory.getContents()); // <-- Here is the violation
```

I am on version 4.9.3
And the issue appears on some machines, but on others it does not appear.
So I havent the issue on one machine with `Temurin-21.0.6+7` while on an other machine there is no issue and I get the issue every time with zulu
